File: backend/database/scripts/import-room-samples.py
#!/usr/bin/env python3
"""
Import room samples data to AnalysisDataset and AnalysisReadyData tables.

This script:
1. Reads *.json files to create AnalysisDataset records
2. Reads *.csv files to populate AnalysisReadyData linked to datasets
3. Uses rooms_metadata.csv for building/floor/occupant info

Usage:
    python scripts/import-room-samples.py
"""

import sqlite3
import json
import csv
import os
import sys
from datetime import datetime
from pathlib import Path

# Configuration
ROOM_SAMPLES_DIR = Path(__file__).parent.parent.parent / "preprocessing" / "room_samples_for_pu"
DB_PATH = Path(__file__).parent.parent / "prisma" / "pu_practice.db"

def generate_cuid():
    """Generate a simple CUID-like ID"""
    import random
    import string
    timestamp = hex(int(datetime.now().timestamp() * 1000))[2:]
    random_part = ''.join(random.choices(string.ascii_lowercase + string.digits, k=8))
    return f"c{timestamp}{random_part}"

def load_metadata():
    """Load room metadata from CSV"""
    metadata_file = ROOM_SAMPLES_DIR / "rooms_metadata.csv"
    metadata = {}

    with open(metadata_file, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            metadata[row['room_id']] = {
                'building': row['building'],
                'floor': row['floor'],
                'room': row['room'],
                'occupant_type': row['occupant_type'],
                'l1_device': row['l1_device'],
                'l2_device': row['l2_device']
            }

    return metadata

def import_analysis_datasets(conn, metadata):
    """Import AnalysisDataset records from JSON summaries"""
    cursor = conn.cursor()
    dataset_mapping = {}

    # Find all JSON summary files
    json_files = list(ROOM_SAMPLES_DIR.glob("room_summary_*.json"))

    print(f"Found {len(json_files)} JSON summary files")

    for json_file in json_files:
        room_id = json_file.stem.replace("room_summary_", "")

        if room_id not in metadata:
            print(f"Warning: No metadata found for {room_id}, skipping")
            continue

        # Load JSON summary
        with open(json_file, 'r', encoding='utf-8') as f:
            summary = json.load(f)

        room_meta = metadata[room_id]
        data_summary = summary['data_summary']

        # Generate dataset ID
        dataset_id = generate_cuid()

        # Parse time range
        start_time = datetime.fromisoformat(data_summary['time_range']['start'])
        end_time = datetime.fromisoformat(data_summary['time_range']['end'])

        # Create dataset record
        cursor.execute("""
            INSERT INTO analysis_datasets (
                id, name, description, building, floor, room,
                start_date, end_date, occupant_type,
                meter_id_l1, meter_id_l2, total_records, positive_labels,
                created_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            dataset_id,
            f"Room {room_id} Analysis Dataset",
            f"Preprocessed analysis data for {room_id} - {data_summary['total_samples']} samples",
            room_meta['building'],
            room_meta['floor'],
            room_meta['room'],
            start_time.isoformat(),
            end_time.isoformat(),
            room_meta['occupant_type'],
            room_meta['l1_device'],
            room_meta['l2_device'],
            data_summary['total_samples'],
            data_summary['positive_samples'],
            datetime.now().isoformat()
        ))

        dataset_mapping[room_id] = dataset_id
        print(f"Created dataset for {room_id}: {dataset_id}")

    conn.commit()
    return dataset_mapping

def import_analysis_ready_data(conn, dataset_mapping):
    """Import AnalysisReadyData records from CSV files"""
    cursor = conn.cursor()

    # Find all CSV sample files
    csv_files = list(ROOM_SAMPLES_DIR.glob("room_samples_*.csv"))

    print(f"Found {len(csv_files)} CSV sample files")

    total_imported = 0

    for csv_file in csv_files:
        room_id = csv_file.stem.replace("room_samples_", "")

        if room_id not in dataset_mapping:
            print(f"Warning: No dataset found for {room_id}, skipping CSV import")
            continue

        dataset_id = dataset_mapping[room_id]

        print(f"Importing CSV data for {room_id}...")

        with open(csv_file, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            batch_data = []
            batch_size = 1000

            for row in reader:
                # Generate record ID
                record_id = generate_cuid()

                # Parse timestamp
                timestamp = datetime.fromisoformat(row['timestamp'])

                # Extract core fields (matching schema)
                source_anomaly_event_id = row['sourceAnomalyEventId'].strip() if row['sourceAnomalyEventId'] and row['sourceAnomalyEventId'].strip() else None

                batch_data.append((
                    record_id,
                    dataset_id,
                    timestamp.isoformat(),
                    row['room_id'],
                    float(row['rawWattageL1']) if row['rawWattageL1'] else 0.0,
                    float(row['rawWattageL2']) if row['rawWattageL2'] else 0.0,
                    float(row['wattage110v_current']) if row['wattage110v_current'] else 0.0,
                    float(row['wattage220v_current']) if row['wattage220v_current'] else 0.0,
                    float(row['wattageTotal_current']) if row['wattageTotal_current'] else 0.0,
                    row['isPositiveLabel'].lower() == 'true',
                    source_anomaly_event_id
                ))

                # Insert in batches
                if len(batch_data) >= batch_size:
                    cursor.executemany("""
                        INSERT INTO analysis_ready_data (
                            id, dataset_id, timestamp, room,
                            raw_wattage_l1, raw_wattage_l2,
                            wattage_110v, wattage_220v, wattage_total,
                            is_positive_label, source_anomaly_event_id
                        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """, batch_data)

                    total_imported += len(batch_data)
                    print(f"  Imported {total_imported} records...")
                    batch_data = []

            # Insert remaining records
            if batch_data:
                cursor.executemany("""
                    INSERT INTO analysis_ready_data (
                        id, dataset_id, timestamp, room,
                        raw_wattage_l1, raw_wattage_l2,
                        wattage_110v, wattage_220v, wattage_total,
                        is_positive_label, source_anomaly_event_id
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """, batch_data)

                total_imported += len(batch_data)

        print(f"Completed {room_id}: {total_imported} total records")

    conn.commit()
    return total_imported

def main():
    """Main import process"""
    print("Starting room samples import...")
    print(f"Source directory: {ROOM_SAMPLES_DIR}")
    print(f"Database path: {DB_PATH}")

    # Check if source directory exists
    if not ROOM_SAMPLES_DIR.exists():
        print(f"Error: Source directory not found: {ROOM_SAMPLES_DIR}")
        sys.exit(1)

    # Check if database exists
    if not DB_PATH.exists():
        print(f"Error: Database not found: {DB_PATH}")
        print("Please run 'pnpm db:reset' first to create the database")
        sys.exit(1)

    # Load metadata
    print("Loading room metadata...")
    metadata = load_metadata()
    print(f"Loaded metadata for {len(metadata)} rooms")

    # Connect to database
    conn = sqlite3.connect(DB_PATH)

    try:
        # Import datasets
        print("\n=== Importing AnalysisDataset records ===")
        dataset_mapping = import_analysis_datasets(conn, metadata)

        # Import ready data
        print("\n=== Importing AnalysisReadyData records ===")
        total_records = import_analysis_ready_data(conn, dataset_mapping)

        print(f"\n=== Import completed ===")
        print(f"Created {len(dataset_mapping)} datasets")
        print(f"Imported {total_records} analysis ready data records")

    except Exception as e:
        print(f"Error during import: {e}")
        conn.rollback()
        sys.exit(1)

    finally:
        conn.close()

if __name__ == "__main__":
    main()

File: backend/database/scripts/import-room-samples.js
/**
 * 房間樣本數據匯入腳本
 *
 * 功能：
 * 1. 讀取 *.json 檔案創建 AnalysisDataset 記錄
 * 2. 讀取 *.csv 檔案填充 AnalysisReadyData 記錄
 * 3. 使用 rooms_metadata.csv 提供建築物/樓層資訊
 */

const { PrismaClient } = require("../prisma/generated");
const fs = require("fs");
const path = require("path");
const csv = require("csv-parser");

const prisma = new PrismaClient();

// ========== 配置 ==========
const ROOM_SAMPLES_DIR = path.join(
	__dirname,
	"../../preprocessing/room_samples_for_pu",
);
const METADATA_FILE = path.join(ROOM_SAMPLES_DIR, "rooms_metadata.csv");

/**
 * 讀取並解析房間元資料
 */
async function loadRoomMetadata() {
	return new Promise((resolve, reject) => {
		const metadata = {};

		if (!fs.existsSync(METADATA_FILE)) {
			reject(new Error(`元資料文件不存在: ${METADATA_FILE}`));
			return;
		}

		fs.createReadStream(METADATA_FILE)
			.pipe(csv())
			.on("data", (data) => {
				metadata[data.room_id] = {
					building: data.building,
					floor: data.floor,
					room: data.room,
					occupantType: data.occupant_type || "OFFICE_WORKER",
					l1Device: data.l1_device,
					l2Device: data.l2_device,
					isHighQuality: data.is_high_quality === "True",
				};
			})
			.on("end", () => {
				console.log(
					`✅ 已讀取 ${Object.keys(metadata).length} 筆房間元資料`,
				);
				resolve(metadata);
			})
			.on("error", reject);
	});
}

/**
 * 讀取房間樣本 JSON 檔案來建立 AnalysisDataset
 */
async function loadDatasetFromJson(jsonFilePath) {
	if (!fs.existsSync(jsonFilePath)) {
		throw new Error(`JSON 檔案不存在: ${jsonFilePath}`);
	}

	const jsonContent = JSON.parse(fs.readFileSync(jsonFilePath, "utf8"));

	return {
		roomId: jsonContent.room_id,
		startDate: jsonContent.data_summary.time_range.start,
		endDate: jsonContent.data_summary.time_range.end,
		totalRecords: jsonContent.data_summary.total_samples,
		positiveLabels: jsonContent.data_summary.positive_samples,
	};
}

/**
 * 讀取房間樣本 CSV 文件
 */
async function loadRoomSampleData(csvFilePath) {
	return new Promise((resolve, reject) => {
		const data = [];

		if (!fs.existsSync(csvFilePath)) {
			reject(new Error(`CSV 文件不存在: ${csvFilePath}`));
			return;
		}

		fs.createReadStream(csvFilePath)
			.pipe(csv())
			.on("data", (row) => {
				// 標準化時間戳格式 (移除毫秒和時區)
				let timestamp = row.timestamp;
				if (timestamp) {
					// 移除毫秒部分 (.000) 和時區部分 (+00:00)
					timestamp = timestamp.replace(
						/\.\d{3}[+-]\d{2}:\d{2}$/,
						"",
					);
					timestamp = timestamp.replace(/[+-]\d{2}:\d{2}$/, "");
				}

				data.push({
					timestamp: new Date(timestamp),
					room: row.room_id,
					rawWattageL1: Number.parseFloat(row.rawWattageL1) || 0,
					rawWattageL2: Number.parseFloat(row.rawWattageL2) || 0,
					wattage110v:
						Number.parseFloat(row.wattage110v_current) || 0,
					wattage220v:
						Number.parseFloat(row.wattage220v_current) || 0,
					wattageTotal:
						Number.parseFloat(row.wattageTotal_current) || 0,
					isPositiveLabel:
						row.isPositiveLabel === "True" ||
						row.isPositiveLabel === true,
					sourceAnomalyEventId:
						row.sourceAnomalyEventId &&
						row.sourceAnomalyEventId !== ""
							? row.sourceAnomalyEventId
							: null,
				});
			})
			.on("end", () => {
				console.log(
					`✅ 從 ${csvFilePath} 讀取了 ${data.length} 筆記錄`,
				);
				resolve(data);
			})
			.on("error", reject);
	});
}

/**
 * 創建 AnalysisDataset 記錄
 */
async function createAnalysisDataset(metadata) {
	const dataset = await prisma.analysisDataset.create({
		data: {
			name: `${metadata.building}-${metadata.floor}-${metadata.room}-Golden-Week`,
			description: `Golden Week 數據集 - ${metadata.building} ${metadata.floor} ${metadata.room}`,
			building: metadata.building,
			floor: metadata.floor,
			room: metadata.room,
			startDate: new Date(metadata.startDate),
			endDate: new Date(metadata.endDate),
			occupantType: metadata.occupantType,
			meterIdL1: metadata.l1Device || "",
			meterIdL2: metadata.l2Device || "",
			totalRecords: metadata.totalRecords,
			positiveLabels: metadata.positiveLabels,
		},
	});

	console.log(`✅ 已創建數據集: ${dataset.name}`);
	return dataset;
}

/**
 * 批量插入 AnalysisReadyData
 */
async function insertAnalysisReadyData(datasetId, samples) {
	const BATCH_SIZE = 1000;
	const totalBatches = Math.ceil(samples.length / BATCH_SIZE);

	console.log(
		`📊 開始批量插入 ${samples.length} 筆數據，分 ${totalBatches} 批次...`,
	);

	for (let i = 0; i < totalBatches; i++) {
		const start = i * BATCH_SIZE;
		const end = Math.min((i + 1) * BATCH_SIZE, samples.length);
		const batch = samples.slice(start, end);

		const dataToInsert = batch.map((sample) => ({
			datasetId: datasetId,
			timestamp: sample.timestamp,
			room: sample.room,
			rawWattageL1: sample.rawWattageL1,
			rawWattageL2: sample.rawWattageL2,
			wattage110v: sample.wattage110v,
			wattage220v: sample.wattage220v,
			wattageTotal: sample.wattageTotal,
			isPositiveLabel: sample.isPositiveLabel,
			sourceAnomalyEventId: sample.sourceAnomalyEventId,
		}));

		await prisma.analysisReadyData.createMany({
			data: dataToInsert,
		});

		console.log(
			`✅ 已完成第 ${i + 1}/${totalBatches} 批次 (${end}/${samples.length} 筆)`,
		);
	}
}

/**
 * 主要匯入流程
 */
async function main() {
	try {
		console.log("🚀 開始匯入房間樣本數據...");

		// 1. 讀取房間元資料
		const roomMetadataMap = await loadRoomMetadata();
		const roomIds = Object.keys(roomMetadataMap);

		// 2. 清理現有數據（可選）
		console.log("🧹 清理現有數據...");
		await prisma.analysisReadyData.deleteMany();
		await prisma.analysisDataset.deleteMany();

		// 3. 逐個處理房間
		for (let i = 0; i < roomIds.length; i++) {
			const roomId = roomIds[i];
			const metadata = roomMetadataMap[roomId];
			console.log(`\n📂 處理房間 ${i + 1}/${roomIds.length}: ${roomId}`);

			try {
				// 檢查 JSON 和 CSV 檔案是否存在
				const jsonPath = path.join(
					ROOM_SAMPLES_DIR,
					`room_summary_${roomId}.json`,
				);
				const csvPath = path.join(
					ROOM_SAMPLES_DIR,
					`room_samples_${roomId}.csv`,
				);

				if (!fs.existsSync(jsonPath) || !fs.existsSync(csvPath)) {
					console.log(`⚠️  房間 ${roomId} 缺少必要檔案，跳過`);
					continue;
				}

				// 讀取 JSON 資料用於建立 AnalysisDataset
				const datasetInfo = await loadDatasetFromJson(jsonPath);

				// 創建數據集記錄，結合 metadata 和 JSON 資料
				const dataset = await createAnalysisDataset({
					...datasetInfo,
					...metadata,
					roomId: roomId,
				});

				// 讀取樣本數據
				const samples = await loadRoomSampleData(csvPath);

				if (samples.length > 0) {
					// 插入樣本數據
					await insertAnalysisReadyData(dataset.id, samples);
					console.log(`✅ 房間 ${roomId} 處理完成`);
				} else {
					console.log(`⚠️  房間 ${roomId} 無有效樣本數據`);
				}
			} catch (error) {
				console.error(
					`❌ 處理房間 ${roomId} 時發生錯誤:`,
					error.message,
				);
				// 繼續處理下一個房間
			}
		}

		// 4. 顯示匯入統計
		const datasetCount = await prisma.analysisDataset.count();
		const dataCount = await prisma.analysisReadyData.count();

		console.log("\n🎉 匯入完成！");
		console.log("📊 統計資訊:");
		console.log(`   - 數據集數量: ${datasetCount}`);
		console.log(`   - 樣本數據總數: ${dataCount}`);
	} catch (error) {
		console.error("❌ 匯入過程中發生錯誤:", error);
		process.exit(1);
	} finally {
		await prisma.$disconnect();
	}
}

// 執行匯入
if (require.main === module) {
	main().catch(console.error);
}

module.exports = { main, loadRoomMetadata, loadRoomSampleData };
